webapp/src/monaco.tsx

```diff
@@ -1271,3 +1202,406 @@ export class Editor extends srceditor.Editor {
+            return names
+                .sort(([, md1], [, md2]) => {
+                    // sort by fn weight
+                    const w2 = (md2 ? md2.weight || 50 : 50);
+                    const w1 = (md1 ? md1.weight || 50 : 50);
+                    return w2 - w1;
+                }).map(([ns, md]) => {
+                    if (!snippets.isBuiltin(ns)) {
+                        const blocks = parent.nsMap[ns].filter(block => !(block.attributes.blockHidden || block.attributes.deprecated));
+                        if (!filterCategory(ns, blocks)) return;
+                        const categoryName = md.block ? md.block : undefined;
+                        return {
+                            ns: ns,
+                            category: categoryName,
+                            color: md.color,
+                            icon: md.icon,
+                            groups: md.groups,
+                            labelLineWidth: md.labelLineWidth,
+                            injectIconClass: true
+                        }
+                    }
+                    else {
+                        const cat = snippets.getBuiltinCategory(ns);
+                        let blocks = cat.blocks || [];
+                        const categoryName = cat.name;
+                        blocks.forEach(b => { b.noNamespace = true })
+                        if (!cat.custom && parent.nsMap[ns.toLowerCase()]) blocks = blocks.concat(parent.nsMap[ns.toLowerCase()].filter(block => !(block.attributes.blockHidden || block.attributes.deprecated)));
+                        if (!blocks || !blocks.length) return;
+                        if (!filterCategory(ns, blocks)) return;
+                        return {
+                            ns: ns,
+                            category: categoryName,
+                            color: md.color,
+                            icon: md.icon,
+                            groups: md.groups,
+                            labelLineWidth: md.labelLineWidth
+                        }
+                    }
+                });
+        }
+
+        const hasAdvanced = namespaces.some(([, md]) => md.advanced);
+        const hasPackages = !pxt.shell.isReadOnly() && pxt.appTarget.cloud && pxt.appTarget.cloud.packages;
+
+        let nonAdvancedCategories = createCategories(namespaces.filter(([, md]) => !(md.advanced)));
+        let advancedCategories = showAdvanced && hasAdvanced ? createCategories(namespaces.filter(([, md]) => md.advanced)) : [];
+
+        function stitchCategories(categories: TreeRowProperties[]): TreeRowProperties[] {
+            // Setup previous and next pointers for keyboard navigation between elements
+            for (let i = 0; i < categories.length; i++) {
+                const categoryInfo = categories[i];
+                if (i > 0) {
+                    (categories[i] as any).previousTreeRow = categories[i - 1];;
```

Comment:
Let's add these two properties to the interface. Just mark it as `/*@internal*/`